src/qibolab/instruments/emulator/backends/qutip_backend.py

```diff
@@ -0,0 +1,586 @@
+        reordering_flip = np.append(flipped_q_indices, flipped_c_indices).astype(int)
+
+        return state.permute(reordering_flip)
+
+    def qevolve(
+        self,
+        channel_waveforms: dict,
+        simulate_dissipation: bool = False,
+    ) -> tuple[np.ndarray, List[int]]:
+        """Performs the quantum dynamics simulation.
+
+        Args:
+            channel_waveforms (dict): The dictionary containing the list of discretized time steps and the corresponding channel waveform amplitudes labelled by the respective channel names.
+            simulate_dissipation (bool): Flag to add (True) or not (False) the dissipation terms associated with T1 and T2 times.
+
+        Returns:
+            tuple: A tuple containing the reduced density matrix of the quantum state at the end of simulation in the Hilbert space specified by the qubits present in the readout channels (little endian), as well as the corresponding list of qubit indices.
+        """
+        full_time_list = channel_waveforms["time"]
+        channel_names = list(channel_waveforms["channels"].keys())
+
+        fp_list = []
+        for channel_name in channel_names:
+            fp_list.append(
+                function_from_array(
+                    channel_waveforms["channels"][channel_name], full_time_list
+                )
+            )
+
+        drift = self.drift
+        scheduled_operators = []
+        ro_qubit_list = []
+
+        # add corresponding operators for non-readout channels to scheduled_operators; add qubit indices of readout channels to ro_qubit_list
+        for channel_name in channel_names:
+            if channel_name[:2] != "R-":
+                scheduled_operators.append(self.operators[channel_name])
+            else:
+                ro_qubit_list.append(int(channel_name[2:]))
+        ro_qubit_list = np.flip(np.sort(ro_qubit_list))
+
+        if simulate_dissipation is True:
+            static_dissipators = self.static_dissipators
+        else:
+            static_dissipators = []
+
+        H = [drift]
+        for i, op in enumerate(scheduled_operators):
+            H.append([op, fp_list[i]])
+        self.H.append(H)
+
+        sim_start_time = timer()
+        if self.sim_method == "master_equation":
+            result = mesolve(
+                H,
+                self.psi0,
+                full_time_list,
+                c_ops=static_dissipators,
+                options=self.sim_opts,
+                progress_bar=EnhancedTextProgressBar(
+                    len(full_time_list), int(len(full_time_list) / 100)
+                ),
+            )
+
+        sim_end_time = timer()
+        sim_time = sim_end_time - sim_start_time
+
+        final_state = result.states[-1]
+        # saves history of states (in little endian, opposite to qibo convention)
+        self.pulse_sim_history.append(result.states)
+        self.pulse_sim_time_list.append(full_time_list)
+        self.pulse_sim_time.append(sim_time)
```

Comment:
Do not store all these values in the object itself. Instead, bundle them and return.

Moreover, in the `PulseSimulator` you will have to satisfy the `Controller` interface, and return a `Result` object, so there would be no place to return this information.

My personal advice would be to subclass the suitable `Result`, and add this information to that object.

It should be the user responsibility to accumulate and manage the history (though, if very relevant for all the emulator operations, you could provide an object and facilities to do that), but the *instrument* is not expected to retain the execution results after download.

Reply:
It seems like there is no need to subclass `Result`, as the `Result` objects are bundled together in a dict by `play` anyway. So I would just add to that dict a 'simulation' key whose value is a dict containing all the simulation related outputs/results.

Reply:
I'm not sure it is an optimal solution, but I don't have a strong motivation for an alternative, so keep going :)

Reply:
In this respect, you're perfectly right.

The weird part of the dictionary is that usually is a mapping `measurement: result`, and this would a non-uniform element.

However, it is true that it will be just ignored, since I expect no one is iterating on the keys, and try to match the given measurement (but instead just accessing the dict for the measurement you're interested in). Though this has to be confirmed.

In any case, let's even start like this, and then (if ever) we could look for some other escape hatch to return this object (even extending the `play` return to a tuple or something like). But one step at a time.

Reply:
> It seems like there is no need to subclass `Result`, as the `Result` objects are bundled together in a dict by `play` anyway. So I would just add to that dict a 'simulation' key whose value is a dict containing all the simulation related outputs/results.

Implemented in d8b2ebd32e5d0bc392ee02eab54875ac842500ab

Reply:
Although I should note that when executing qibo circuits, because `execute_circuit` doesn't actually return the same object as `execute_pulse_sequence` but instead extracts information from there and puts them in `MeasurementOutcomes`, the additional simulation related data I added to the output of `execute_pulse_sequence` doesn't carry over to `execute_circuit`. Nonetheless it runs and will simply return what all other qibolab platforms return and nothing more.

Reply:
No problem for Qibo: running circuits through the backend is the base user behavior, and you only usually care about the results.

But it's always possible to manually compile the circuit, and execute the pulse sequence. If really needed, for the time being it will be accessed this way. We're always on time to extend the interface later on (if breaking it would be already breaking now, so better avoid, and in case sync with a major release, and if not, we could do it at any time).